### PR TITLE
Fix multiple CUDA BatchNorm bugs

### PR DESCRIPTION
**Pull request by mtavenrath**
### Description
- Fix incorrect running_mean / running_var in training mode due to incorrect momentum and missing input mean/var. runnig_var could be correct, but has a too high epsilon.
- Fix incorrect checks when using NHWC
- Pass NHWC flag to NormalizeDims to get correct new dimensions from x_shape
- Register missing double operations to get parity between NHWC/NCHW

### CODE REVIEW

File: onnxruntime/core/providers/cuda/nn/batch_norm.cc [hunk 1]
```diff
@@ -137,6 +137,10 @@ Status BatchNorm<T, NHWC>::ComputeInternal(OpKernelContext* p_op_kernel_context)
     auto saved_mean_data = reinterpret_cast<CudaT*>(saved_mean->MutableData<T>());
     auto saved_inv_var_data = reinterpret_cast<CudaT*>(saved_var->MutableData<T>());
 
+    auto stream = static_cast<cudaStream_t>(p_op_kernel_context->GetComputeStream()->GetHandle());
+    cudaMemcpyAsync(running_mean_data, mean_data, mean->SizeInBytes(), cudaMemcpyDeviceToDevice, stream);
```
Comment (tianleiwu) [inline onnxruntime/core/providers/cuda/nn/batch_norm.cc:141]: use CUDA_RETURN_IF_ERROR(...) to check return value of cudaMemcpyAsync

File: onnxruntime/core/providers/cuda/nn/batch_norm.cc [hunk 2]
```diff
@@ -149,7 +155,7 @@ Status BatchNorm<T, NHWC>::ComputeInternal(OpKernelContext* p_op_kernel_context)
         bn_tensor_desc,
         scale_data,
         b_data,
-        momentum_,
+        1.0 - momentum_,
```
Comment (tianleiwu) [inline onnxruntime/core/providers/cuda/nn/batch_norm.cc:158]: Good catch.

I found a a blog for the difference between momentum and exponentialAverageFactor: https://kaixih.github.io/batch-norm/

**Comment by tianleiwu** [comment]
@mtavenrath, can we remove the check here?
https://github.com/microsoft/onnxruntime/blob/70cd96684d5874faef5de492925dde4ce02a29a6/onnxruntime/core/providers/cuda/nn/batch_norm.h#L44

File: onnxruntime/core/providers/cuda/cuda_execution_provider.cc [hunk 3]
```diff
@@ -1202,9 +1202,9 @@ class ONNX_OPERATOR_TYPED_KERNEL_CLASS_NAME(kCudaExecutionProvider, kOnnxDomain,
 class ONNX_OPERATOR_TYPED_KERNEL_CLASS_NAME(kCudaExecutionProvider, kOnnxDomain, 14, float, LSTM);
 class ONNX_OPERATOR_TYPED_KERNEL_CLASS_NAME(kCudaExecutionProvider, kOnnxDomain, 14, double, LSTM);
 class ONNX_OPERATOR_TYPED_KERNEL_CLASS_NAME(kCudaExecutionProvider, kOnnxDomain, 14, MLFloat16, LSTM);
-class ONNX_OPERATOR_VERSIONED_TYPED_KERNEL_CLASS_NAME(kCudaExecutionProvider, kOnnxDomain, 14, 14, float, BatchNormalization);
-class ONNX_OPERATOR_VERSIONED_TYPED_KERNEL_CLASS_NAME(kCudaExecutionProvider, kOnnxDomain, 14, 14, double, BatchNormalization);
-class ONNX_OPERATOR_VERSIONED_TYPED_KERNEL_CLASS_NAME(kCudaExecutionProvider, kOnnxDomain, 14, 14, MLFloat16, BatchNormalization);
+class ONNX_OPERATOR_VERSIONED_KERNEL_CLASS_NAME(kCudaExecutionProvider, kOnnxDomain, 14, 14, float, BatchNormalization);
```
Comment (tianleiwu) [inline onnxruntime/core/providers/cuda/cuda_execution_provider.cc:1205]: ONNX_OPERATOR_VERSIONED_TYPED_KERNEL_CLASS_NAME

File: onnxruntime/core/providers/cuda/cuda_execution_provider.cc [hunk 4]
```diff
@@ -2107,9 +2107,12 @@ static Status RegisterCudaKernels(KernelRegistry& kernel_registry) {
     BuildKernelCreateInfo<ONNX_OPERATOR_TYPED_KERNEL_CLASS_NAME(kCudaExecutionProvider, kOnnxDomain, 14, double, LSTM)>,
     BuildKernelCreateInfo<ONNX_OPERATOR_TYPED_KERNEL_CLASS_NAME(kCudaExecutionProvider, kOnnxDomain, 14, MLFloat16, LSTM)>,
     BuildKernelCreateInfo<ONNX_OPERATOR_VERSIONED_KERNEL_CLASS_NAME(kCudaExecutionProvider, kOnnxDomain, 14, 18, Reshape)>,
-    BuildKernelCreateInfo<ONNX_OPERATOR_VERSIONED_TYPED_KERNEL_CLASS_NAME(kCudaExecutionProvider, kOnnxDomain, 14, 14, float, BatchNormalization)>,
-    BuildKernelCreateInfo<ONNX_OPERATOR_VERSIONED_TYPED_KERNEL_CLASS_NAME(kCudaExecutionProvider, kOnnxDomain, 14, 14, double, BatchNormalization)>,
-    BuildKernelCreateInfo<ONNX_OPERATOR_VERSIONED_TYPED_KERNEL_CLASS_NAME(kCudaExecutionProvider, kOnnxDomain, 14, 14, MLFloat16, BatchNormalization)>,
+    BuildKernelCreateInfo<ONNX_OPERATOR_VERSIONED_KERNEL_CLASS_NAME(
```
Comment (tianleiwu) [inline onnxruntime/core/providers/cuda/cuda_execution_provider.cc:2110]: ONNX_OPERATOR_VERSIONED_TYPED_KERNEL_CLASS_NAME

File: onnxruntime/core/providers/cuda/cuda_execution_provider.cc [hunk 5]
```diff
@@ -1202,9 +1202,12 @@ class ONNX_OPERATOR_TYPED_KERNEL_CLASS_NAME(kCudaExecutionProvider, kOnnxDomain,
 class ONNX_OPERATOR_TYPED_KERNEL_CLASS_NAME(kCudaExecutionProvider, kOnnxDomain, 14, float, LSTM);
 class ONNX_OPERATOR_TYPED_KERNEL_CLASS_NAME(kCudaExecutionProvider, kOnnxDomain, 14, double, LSTM);
 class ONNX_OPERATOR_TYPED_KERNEL_CLASS_NAME(kCudaExecutionProvider, kOnnxDomain, 14, MLFloat16, LSTM);
-class ONNX_OPERATOR_VERSIONED_TYPED_KERNEL_CLASS_NAME(kCudaExecutionProvider, kOnnxDomain, 14, 14, float, BatchNormalization);
-class ONNX_OPERATOR_VERSIONED_TYPED_KERNEL_CLASS_NAME(kCudaExecutionProvider, kOnnxDomain, 14, 14, double, BatchNormalization);
-class ONNX_OPERATOR_VERSIONED_TYPED_KERNEL_CLASS_NAME(kCudaExecutionProvider, kOnnxDomain, 14, 14, MLFloat16, BatchNormalization);
+class ONNX_OPERATOR_VERSIONED_TYPED_KERNEL_CLASS_NAME(
```
Comment (hariharans29) [inline onnxruntime/core/providers/cuda/cuda_execution_provider.cc:1205]: I think the support for type `double` for the NCHW `BatchNorm` came with the very first commit. Since then the strategy has changed to be more careful in adding further type support unless warranted by a production use-case. Even though currently the CUDA NHWC support is turned OFF by default, if and when we turn it ON by default - we may have to deal with all the excess binary size growth. Maybe we should hold off on supporting the `double` type for NHWC `BatchNorm` for now if the only reason for that enhancement is to get type parity with NCHW `BatchNorm` ?

Comment (tianleiwu) [reply onnxruntime/core/providers/cuda/cuda_execution_provider.cc:1205]: We can add add a build flag later to enable/disable all double NHWC ops.

**Review by tianleiwu — APPROVED** [review]
